chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -1330,39 +1452,123 @@ impl<'a, S: BlockchainStorageWrite, V: TransactionVerificationStrategy> Chainsta
     #[log_error]
-    pub fn del_block_index_of_non_persisted_block(
+    pub fn del_block_indices_of_non_persisted_block_tree<'b>(
         &mut self,
-        block_id: &Id<Block>,
+        tree: InMemoryBlockTreeRef<'b>,
     ) -> Result<(), BlockError> {
-        if let Some(existing_block_index) = self.db_tx.get_block_index(block_id)? {
+        let old_leaf_block_ids = self.db_tx.get_leaf_block_ids(
+            tree.root_block_index()
+                .block_height()
+                .prev_height()
+                .expect("Non-genesis can't have zero height"),
+        )?;
+        let mut remaining_leaf_block_ids = old_leaf_block_ids.clone();
+
+        for block_index in tree.iter_all_block_indices() {
+            let block_id = *block_index.block_id();
+
             // Note: here we're being extra-cautious about someone mis-using this function, so we only panic in
             // debug mode.
             debug_assert_or_log!(
-                !existing_block_index.is_persisted(),
-                "Trying to delete a block index for a persisted block {block_id}"
+                !block_index.is_persisted(),
+                "Trying to delete a block index for a persisted block {block_id}",
             );
 
-            self.db_tx.del_block_index(*block_id)?;
+            self.db_tx.del_block_index(block_id)?;
+
+            if remaining_leaf_block_ids.remove(&block_id) {
+                self.db_tx.mark_as_leaf(block_index, false)?;
+            }
         }
+
+        // The parent of the removed subtree's root may have become a new leaf.
+        // Check if it's a parent of one of the remaining leaves; if not, mark it as a leaf.
+        let maybe_new_leaf_id = tree.root_block_index().prev_block_id();
+        if let Some(maybe_new_leaf_id) =
+            maybe_new_leaf_id.classify(self.chain_config).chain_block_id()
+        {
+            let mut is_parent = false;
+
+            for leaf_block_id in remaining_leaf_block_ids {
+                let leaf_block_index = self
+                    .get_existing_block_index(&leaf_block_id)
+                    .map_err(BlockError::PropertyQueryError)?;
+
+                if self
+                    .is_parent_of(&leaf_block_index, &maybe_new_leaf_id)
+                    .map_err(BlockError::PropertyQueryError)?
+                {
+                    is_parent = true;
+                    break;
+                }
+            }
+
+            if !is_parent {
+                let maybe_new_leaf_index = self
+                    .get_existing_block_index(&maybe_new_leaf_id)
+                    .map_err(BlockError::PropertyQueryError)?;
+                self.db_tx.mark_as_leaf(&maybe_new_leaf_index, true)?;
+            }
+        }
+
         Ok(())
     }
 
+    /// Check if the block specified by `potential_parent_id` is a (possibly indirect) parent of
+    /// the block specified by `child_index`.
+    #[log_error]
+    fn is_parent_of(
```

Comment:
Question, why not use the function `last_common_ancestor` for the implementation here? 

If not... Note that this looping can be optimized by using "some_ancestor" in BlockIndex. It was used previously.